File: php/src/Number.php
<?php
namespace Vanni\Idmix;

/**
 * 类型转换层：将调用方值规范化为内部 DataObject 表示。
 */
final class Number
{
    public const MAX_STRING_LEN = 63;

    private function __construct()
    {
    }

    /** @param mixed[] $values @return DataObject[] */
    public static function normalizeObjects(array $values): array
    {
        $out = [];
        foreach ($values as $i => $v) {
            try {
                $out[] = self::objectFromAny($v);
            } catch (\InvalidArgumentException $e) {
                throw new \InvalidArgumentException("value[$i]: {$e->getMessage()}", 0, $e);
            }
        }
        return $out;
    }

    /** @param DataObject[] $objects @return array<int, TypedValue|string> */
    public static function materializeObjects(array $objects): array
    {
        $out = [];
        foreach ($objects as $i => $obj) {
            if ($obj->isString) {
                $out[] = $obj->str;
                continue;
            }
            try {
                $out[] = self::materializeValue($obj);
            } catch (\InvalidArgumentException $e) {
                throw new \InvalidArgumentException("value[$i]: {$e->getMessage()}", 0, $e);
            }
        }
        return $out;
    }

    public static function objectFromAny(mixed $v): DataObject
    {
        if ($v instanceof TypedValue) {
            return DataObject::integer($v->otype, $v->val);
        }
        if ($v instanceof Bytes) {
            if ($v->data === '') {
                throw new \InvalidArgumentException(
                    'empty byte slice is not allowed (max ' . self::MAX_STRING_LEN . ' bytes)'
                );
            }
            if (strlen($v->data) > self::MAX_STRING_LEN) {
                throw new \InvalidArgumentException(
                    'byte slice length ' . strlen($v->data) . ' exceeds max ' . self::MAX_STRING_LEN
                );
            }
            return DataObject::string($v->data);
        }
        if (is_string($v)) {
            if ($v === '') {
                throw new \InvalidArgumentException(
                    'empty string is not allowed (max ' . self::MAX_STRING_LEN . ' bytes)'
                );
            }
            if (strlen($v) > self::MAX_STRING_LEN) {
                throw new \InvalidArgumentException(
                    'string length ' . strlen($v) . ' exceeds max ' . self::MAX_STRING_LEN
                );
            }
            return DataObject::string($v);
        }
        if (is_int($v)) {
            IntMath::ensureAvailable();
            return DataObject::integer(TypedValue::OTYPE_INT64, (string) $v);
        }
        throw new \InvalidArgumentException(
            'unsupported type ' . get_debug_type($v) . ' (integer or string up to ' . self::MAX_STRING_LEN . ' bytes)'
        );
    }

    private static function materializeValue(DataObject $obj): TypedValue
    {
        return match ($obj->otype) {
            TypedValue::OTYPE_UINT8 => TypedValue::u8($obj->val),
            TypedValue::OTYPE_UINT16 => TypedValue::u16($obj->val),
            TypedValue::OTYPE_UINT32 => TypedValue::u32($obj->val),
            TypedValue::OTYPE_UINT64 => TypedValue::u64($obj->val),
            TypedValue::OTYPE_INT8 => TypedValue::i8($obj->val),
            TypedValue::OTYPE_INT16 => TypedValue::i16($obj->val),
            TypedValue::OTYPE_INT32 => TypedValue::i32($obj->val),
            TypedValue::OTYPE_INT64 => TypedValue::i64($obj->val),
            default => throw new \InvalidArgumentException("invalid otype {$obj->otype}"),
        };
    }
}

File: rust/lib/src/number.rs
use crate::error::IdMixError;
use crate::idx_codec::{DataObject, MAX_STRING_LEN};
use crate::typed_value::{
    OTYPE_INT16, OTYPE_INT32, OTYPE_INT64, OTYPE_INT8, OTYPE_UINT16, OTYPE_UINT32, OTYPE_UINT64,
    OTYPE_UINT8, Value,
};

pub fn normalize_objects(values: &[Value]) -> Result<Vec<DataObject>, IdMixError> {
    values.iter().map(object_from_value).collect()
}

pub fn object_from_value(v: &Value) -> Result<DataObject, IdMixError> {
    match v {
        Value::String(s) => {
            if s.is_empty() {
                return Err(IdMixError::msg(format!(
                    "empty string is not allowed (max {MAX_STRING_LEN} bytes)"
                )));
            }
            if s.len() > MAX_STRING_LEN {
                return Err(IdMixError::msg(format!(
                    "string length {} exceeds max {MAX_STRING_LEN}",
                    s.len()
                )));
            }
            Ok(DataObject::string(s.as_bytes()))
        }
        Value::Bytes(b) => {
            if b.is_empty() {
                return Err(IdMixError::msg(format!(
                    "empty byte slice is not allowed (max {MAX_STRING_LEN} bytes)"
                )));
            }
            if b.len() > MAX_STRING_LEN {
                return Err(IdMixError::msg(format!(
                    "byte slice length {} exceeds max {MAX_STRING_LEN}",
                    b.len()
                )));
            }
            Ok(DataObject::string(b))
        }
        Value::U8(v) => Ok(DataObject::integer(OTYPE_UINT8, *v as i64)),
        Value::U16(v) => Ok(DataObject::integer(OTYPE_UINT16, *v as i64)),
        Value::U32(v) => Ok(DataObject::integer(OTYPE_UINT32, *v as i64)),
        Value::U64(v) => Ok(DataObject::integer(OTYPE_UINT64, *v as i64)),
        Value::I8(v) => Ok(DataObject::integer(OTYPE_INT8, *v as i64)),
        Value::I16(v) => Ok(DataObject::integer(OTYPE_INT16, *v as i64)),
        Value::I32(v) => Ok(DataObject::integer(OTYPE_INT32, *v as i64)),
        Value::I64(v) => Ok(DataObject::integer(OTYPE_INT64, *v)),
    }
}

pub fn materialize_objects(objects: &[DataObject]) -> Result<Vec<Value>, IdMixError> {
    objects.iter().map(materialize_object).collect()
}

fn materialize_object(obj: &DataObject) -> Result<Value, IdMixError> {
    if obj.is_string {
        return Ok(Value::String(
            String::from_utf8(obj.str.clone()).map_err(|e| IdMixError::msg(e.to_string()))?,
        ));
    }
    materialize_value(obj.otype, obj.val)
}

fn materialize_value(otype: u8, val: i64) -> Result<Value, IdMixError> {
    match otype {
        OTYPE_UINT8 => Ok(Value::U8(val as u8)),
        OTYPE_UINT16 => Ok(Value::U16(val as u16)),
        OTYPE_UINT32 => Ok(Value::U32(val as u32)),
        OTYPE_UINT64 => Ok(Value::U64(val as u64)),
        OTYPE_INT8 => Ok(Value::I8(val as i8)),
        OTYPE_INT16 => Ok(Value::I16(val as i16)),
        OTYPE_INT32 => Ok(Value::I32(val as i32)),
        OTYPE_INT64 => Ok(Value::I64(val)),
        _ => Err(IdMixError::msg(format!("invalid otype {otype}"))),
    }
}
